runelite-client/src/main/java/net/runelite/client/ui/ClientUI.java

```diff
@@ -317,14 +326,76 @@ public void init(final RuneLite runelite) throws Exception
 			frame.setLocationRelativeTo(frame.getOwner());
 			frame.setResizable(true);
 
-			SwingUtil.addGracefulExitCallback(frame,
-				() ->
+			frame.setDefaultCloseOperation(JFrame.DO_NOTHING_ON_CLOSE);
+			frame.addWindowListener(new WindowAdapter()
+			{
+				@Override
+				public void windowClosing(WindowEvent event)
 				{
-					saveClientBoundsConfig();
-					runelite.shutdown();
-				},
-				this::showWarningOnExit
-			);
+					int result = JOptionPane.OK_OPTION;
+
+					if (showWarningOnExit())
+					{
+						try
+						{
+							result = JOptionPane.showConfirmDialog(
+								frame,
+								"Are you sure you want to exit?", "Exit",
+								JOptionPane.OK_CANCEL_OPTION,
+								JOptionPane.QUESTION_MESSAGE);
+						}
+						catch (Exception e)
+						{
+							log.warn("Unexpected exception occurred while check for confirm required", e);
+						}
+					}
+
+					if (result == JOptionPane.OK_OPTION)
+					{
+						saveClientBoundsConfig();
+						ClientShutdown csev = new ClientShutdown();
+						eventBus.post(csev);
+						new Thread(() ->
```

Comment:
I dont really like the complexity of this, cant you at least move it to separate method? Its very annoying to read.

Also the method do not needs Runelite as param anymore right